src/NServiceBus.Core.Tests/ObjectBuilder/Common/CommonObjectBuilderTests.cs

```diff
@@ -0,0 +1,181 @@
+        }
+
+        [Test]
+        public void When_using_injected_builder_to_resolve_uow_dependencies()
+        {
+            var builder = new CommonObjectBuilder(new LightInjectObjectBuilder());
+
+            //IBuilder iBuilder = builder;
+            //builder.ConfigureComponent(b => iBuilder, DependencyLifecycle.SingleInstance);
+            builder.ConfigureComponent(_ => builder.CurrentBuilder, DependencyLifecycle.InstancePerCall);
+
+            builder.ConfigureComponent<ClassResolvingItsDependencies>(DependencyLifecycle.InstancePerCall);
+            builder.ConfigureComponent<UowDependency>(DependencyLifecycle.InstancePerUnitOfWork);
+
+            ClassResolvingItsDependencies c2;
+            ClassResolvingItsDependencies c3;
+
+            var c1 = builder.Build<ClassResolvingItsDependencies>();
+
+            using (var childBuilder = builder.CreateChildBuilder())
+            {
+                c2 = childBuilder.Build<ClassResolvingItsDependencies>();
+                c3 = childBuilder.Build<ClassResolvingItsDependencies>();
+            }
+
+            Assert.AreNotSame(c1, c2);
+            Assert.AreNotSame(c2, c3);
+
+            Assert.AreSame(c2.dependency, c3.dependency);
+
+            Assert.AreEqual(2, UowDependency.InstanceCounter);
+            Assert.AreNotSame(c1.dependency, c2.dependency);
+        }
+
+        [Test]
+        public async Task When_using_injected_builder_on_another_thread_to_resolve_uow_dependencies()
+        {
+            var builder = new CommonObjectBuilder(new LightInjectObjectBuilder());
+
+            //IBuilder iBuilder = builder;
+            //builder.ConfigureComponent(b => iBuilder, DependencyLifecycle.SingleInstance);
+            builder.ConfigureComponent(_ => builder.CurrentBuilder, DependencyLifecycle.InstancePerCall);
+
+            builder.ConfigureComponent<ClassResolvingItsDependencies>(DependencyLifecycle.InstancePerCall);
+            builder.ConfigureComponent<UowDependency>(DependencyLifecycle.InstancePerUnitOfWork);
+
+            ClassResolvingItsDependencies c2 = null;
+            ClassResolvingItsDependencies c3 = null;
+
+            var c1 = builder.Build<ClassResolvingItsDependencies>();
+
+            await Task.Run(async () =>
+            {
+                using (var childBuilder = builder.CreateChildBuilder())
+                {
+                    c2 = childBuilder.Build<ClassResolvingItsDependencies>();
+                    await Task.Yield();
+                    c3 = childBuilder.Build<ClassResolvingItsDependencies>();
+                }
+            });
+
+            Assert.AreNotSame(c1, c2);
+            Assert.AreNotSame(c2, c3);
+
+            Assert.AreSame(c2.dependency, c3.dependency);
+
+            Assert.AreEqual(2, UowDependency.InstanceCounter);
+            Assert.AreNotSame(c1.dependency, c2.dependency);
+        }
+
+        [Test]
+        public void When_resolving_uow_dependencies_configured_with_builder_factory()
```

Comment:
I get the feeling this should be two separate tests, it's difficult for me to follow all the asserts to see exactly what this test is checking.

Test name is also missing a conclusion.